README.md
# Opal::Optimizer

A utility to optimize the JavaScript output from Opal with a help of RKelly-Turbo JavaScript parser.

As of now it contains two steps:

* Tree shaking for Opal methods - if a named method is not called anywhere in the code, it's removed
* Collapsing stubs

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'opal-optimizer'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install opal-optimizer

## Usage

For Sprockets environments, all you need is to `require "opal/optimizer/sprockets"` in your pipeline.

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/hmdne/opal-optimizer.


## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
## v0.1.8

* Improve interop with Sprockets when non-Opal javascripts are compiled as well

## v0.1.7

* Opal 1.6 compatibility

## v0.1.6

* Further Opal 1.4 compatibility

## v0.1.5

* Opal 1.4 compatibility


spec/spec_helper.rb
require "bundler/setup"
require "opal/optimizer"
require "opal/optimizer/sprockets"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/opal/optimizer_spec.rb
require "opal-browser"
require "opal/builder"
require "fileutils"

RSpec.describe Opal::Optimizer do
  it "optimizes a basic Opal output" do
    file = Opal::Builder.new.build_str(<<~RUBY, '(opt)').to_s
      require "opal"
    RUBY
    out = Opal::Optimizer.new(file).optimize
    expect(out).not_to be nil
    expect(out).to include "/* destroyed: "
  end

  it "optimizes a basic Opal + opal-browser large output" do
    file = Opal::Builder.new.build_str(<<~RUBY, '(opt)').to_s
      require "opal"
      require "native"
      require "promise"
      require "browser/setup/large"
    RUBY
    out = Opal::Optimizer.new(file).optimize
    expect(out).not_to be nil
    expect(out).to include "/* destroyed: "
  end

  it "optimizes a basic Opal + opal-browser traditional output" do
    file = Opal::Builder.new.build_str(<<~RUBY, '(opt)').to_s
      require "opal"
      require "native"
      require "promise"
      require "browser"
    RUBY
    out = Opal::Optimizer.new(file).optimize
    expect(out).not_to be nil
    expect(out).to include "/* destroyed: "
  end

  it "optimizes an Opal program and allows us to execute it" do
    file = Opal::Builder.new.build_str(<<~RUBY, '(opt)').to_s
      require "opal"
      o = Object.new
      def o.method_missing(meth)
        meth
      end
      puts 5.times.to_a.join
      puts o.hello
    RUBY
    out = Opal::Optimizer.new(file).optimize
    expect(out).not_to be nil
    expect(out).to include "/* destroyed: "
    FileUtils.mkdir_p("tmp")
    File.write("tmp/example.js", out)
    expect(`node tmp/example.js 2>#{Gem.win_platform? ? "NUL" : "/dev/null"}`).to eq("01234\nhello\n")
  end

  # Too large for now, maybe later.
  #
  #it "optimizes a basic Opal 1.1-pre + parser output" do
  #  file = File.read('spec/fixtures/opal-1.1-pre+parser-output.js')
  #  out = Opal::Optimizer.new(file).optimize
  #  expect(out).not_to be nil
  #  expect(out).to include "/* destroyed: "
  #end
end


lib/opal/optimizer.rb
require "opal/optimizer/version"
require "rkelly"
require "opal/optimizer/helpers"
require "opal/compiler"

module Opal
  class Optimizer
    class NonOpalArgumentError < ArgumentError; end

    include Helpers

    attr_accessor :ast, :opal_version, :corelib, :corelib_source, :corelib_calls,
                  :function_calls

    attr_accessor :exports

    def initialize(code, exports: "", force_corelib: true)
      if force_corelib
        unless code.include? "//   A few conventions for the documentation of this file:"
          raise NonOpalArgumentError, "A non-Opal code provided"
        end
      end

      @ast = parse_js(code)

      @corelib = @ast.value.find do |i|
        es = code[i.range.from.index..i.range.to.index]
        if es.start_with?("(function(undefined) {\n  // @note\n"\
                          "  //   A few conventions for the documentation of this file:") &&
           es.end_with?("TypeError.$$super = Error;\n}).call(this);")
          @opal_version = 1.0
        elsif es.start_with?("(function(global_object) {\n  \"use strict\";\n\n  // @note\n  "\
                             "//   A few conventions for the documentation of this file:")
          if es.end_with?("TypeError.$$super = Error;\n}).call(this);")
            if es.include? 'function $prop(object,'
              @opal_version = 1.4
            else
              @opal_version = 1.1
            end
          elsif es.end_with?("Opal.file_sources = {};\n}).call(this);")
            @opal_version = 1.6
          end
        end
      end

      raise ArgumentError, "Couldn't deduce Opal version based on this content" if force_corelib && !@corelib

      @corelib_source = @corelib.value.value.value.value.function_body.value if @corelib

      reload

      # Are exports js or do we need to compile them first?
      unless [nil, ""].include?(exports) || exports.start_with?("(function(")
        exports = Opal::Compiler.new(exports).compile
      end
      @exports = Opal::Optimizer.new(exports, exports: nil, force_corelib: false) unless exports == nil
    end

    def reload
      @function_calls = ast.pointcut(FunctionCallNode).matches
      @corelib_calls = @function_calls.select do |i|
        i.value_path?(DotAccessorNode, ResolveNode, "Opal") ||
        i.value_path?(ResolveNode, /\A\$/)
      end.group_by { |i| (i.value.value.is_a?(String) ? i.value.value : i.value.accessor).gsub('$', '') }
      @corelib_calls = Hash.new { [] }.merge(@corelib_calls)
    end

    def optimize
      Step::TreeShaking.new(self).run
      Step::CollapseStubs.new(self).run

      @ast.to_ecma
    end
  end
end

require "opal/optimizer/step"


lib/opal/optimizer/step/tree_shaking.rb
require "set"

module Opal; class Optimizer; class Step

class TreeShaking < Step
  def shake_methods
    aliases = corelib_calls["alias"].map do |i|
      old = i.arguments.value[2]
      [i, "$"+old.value[1..-2]] if StringNode === old
    end.compact.to_h

    method_defs = corelib_calls["def"] +
                  corelib_calls["defs"] +
                  corelib_calls["defn"] +
                  corelib_calls["udef"] +
                  aliases.keys

    method_calls = Set.new(['$method_missing'])

    method_calls += ( function_calls + exports.function_calls ).map do |i|
      out = if i.value_path?(DotAccessorNode) && i.value.accessor.start_with?("$")
        i.value.accessor
      elsif i.value_path?(BracketAccessorNode) &&
            StringNode === i.value.accessor &&
            i.value.accessor.value[1] == '$'
        i.value.accessor.value[1..-2]
      elsif i.value_path?(ResolveNode, ->(i) { %w[$send $send2 $refined_send].include? i })
        if i.value.value == '$send'
          old = i.arguments.value[1]
        else
          old = i.arguments.value[2]
        end
        "$" + old.value[1..-2] if StringNode === old
      end

      out = [out]

      case out.first
      when /\A\$(public_|private_|protected_)?(class_|instance_|singleton_)?(send|method(_defined\?)?)\z/,
           '$__send__'

        old = i.arguments.value[0]
        out << "$" + old.value[1..-2] if StringNode === old
      end

      out
    end.flatten.compact

    method_calls += aliases.values

    # Protected methods
    method_calls += []

    removed = Set.new

    method_defs.each do |m|
      name = m.arguments.value[1]
      case name
      when AddNode
        if StringNode === name.left && StringNode === name.value
          name = name.left.value[1..-2] + name.value.value[1..-2]
        else
          next
        end
      when StringNode
        name = name.value[1..-2]
        case m.value
        when ResolveNode
          name = "$" + name if m.value.value == "$alias"
        when DotAccessorNode
          name = "$" + name if m.value.accessor == "alias"
        end
      else
        next
      end

      next if method_calls.include? name

      removed << name
      m.destroy! "TreeShaking#shake_methods/#{name}"
    end

    corelib_calls["add_stubs"].each do |stubcall|
      if opal_version >= 1.4
        stubs = stubcall.arguments.value.first.value
        new_stubs = stubs[1..-2].split(",").reject do |i|
          removed.include? i[1..-1]
        end.join(',')
        new_stubs = "'#{new_stubs}'"
        stubs.replace(new_stubs)
      else
        stubcall.arguments.value.first.value.reject! do |i|
          removed.include? i.value.value[1..-2]
        end
      end
    end

    removed
  end

  def run
    loop do
      removed = shake_methods
      # $stdout.puts removed.inspect
      reload
      break if removed.length == 0
    end
  end
end

end; end; end


lib/opal/optimizer/step/collapse_stubs.rb
require 'set'

module Opal; class Optimizer; class Step

class CollapseStubs < Step
  def run
    # We can't collapse when no corelib present.
    return unless corelib

    stubs = Set.new

    nodes = corelib_calls["add_stubs"] || []
    nodes.each do |node|
      if opal_version >= 1.4
        stubs += node.arguments.value.first.value[1..-2].split(',').map(&"$".method(:+))
      else
        stubs += node.arguments.value.first.value.map do |i|
          i.value.value.gsub(/['"]/, '')
        end
      end

      node.destroy! "CollapseStubs"
    end

    stubs -= ["__send__", "class", "!=", "equal?", "==", "__id__", "!",
              "initialize", "eql?", "instance_eval", "instance_exec",
              "singleton_method_added", "singleton_method_removed",
              "singleton_method_undefined", "method_missing"].map(&"$".method(:+))

    stubs -= ["$respond_to_missing?"] if opal_version >= 1.1


    new_stub_code = <<~JAVASCRIPT
      var stubs = '#{stubs.to_a.join(',')}'.split(','), stubs_obj = {};
      for (var i = 0; i < stubs.length; i++)
        stubs_obj[stubs[i]] = {value: Opal.stub_for(stubs[i]), enumerable: false, configurable: true, writable: true};
      Object.defineProperties(Opal.BasicObject.$$prototype, stubs_obj);
    JAVASCRIPT

    new_stub_code = parse_js(new_stub_code)

    corelib_source.value += new_stub_code.value
  end
end

end; end; end
